zephyr/src/object.rs

```diff
@@ -0,0 +1,231 @@
+//! them to APIs in Zephyr that require them.
+//!
+//! Most kernel objects use mutable pointers, and it makes sense to require the wrapper structure
+//! to be mutable to these accesses.  There a few cases, mostly generated ones that live in
+//! read-only memory, notably device instances, that need const pointers.  These will be
+//! represented by a separate wrapper.
+//!
+//! # Initialization tracking
+//!
+//! The Kconfig `CONFIG_RUST_CHECK_KOBJ_INIT` enabled extra checking in Rust-based kernel objects.
+//! This will result in a panic if the objects are used before the underlying object has been
+//! initialized.  The initialization must happen through the `StaticKernelObject::init_help`
+//! method.
+//!
+//! TODO: Document how the wrappers work once we figure out how to implement them.
+
+use core::{cell::UnsafeCell, mem};
+
+use crate::sync::atomic::{AtomicUsize, Ordering};
+
+/// A kernel object represented statically in Rust code.
+///
+/// These should not be declared directly by the user, as they generally need linker decorations to
+/// be properly registered in Zephyr as kernel objects.  The object has the underlying Zephyr type
+/// T, and the wrapper type W.
+///
+/// TODO: Can we avoid the public fields with a const new method?
+///
+/// TODO: Handling const-defined alignment for these.
+pub struct StaticKernelObject<T> {
+    #[allow(dead_code)]
+    /// The underlying zephyr kernel object.
+    pub value: UnsafeCell<T>,
+    /// Initialization status of this object.  Most objects will start uninitialized and be
+    /// initialized manually.
+    pub init: AtomicUsize,
+}
+
+/// Each can be wrapped appropriately.  The wrapped type is the instance that holds the raw pointer.
+pub trait Wrapped {
+    /// The wrapped type.  This is what `take()` on the StaticKernelObject will return after
+    /// initialization.
+    type T;
+
+    /// The wrapped type also requires zero or more initializers.  Which are represented by this
+    /// type.
+    type I;
+
+    /// Initialize this kernel object, and return the wrapped pointer.
+    fn get_wrapped(&self, args: Self::I) -> Self::T;
+}
+
+/// A state indicating an uninitialized kernel object.
+///
+/// This must be zero, as kernel objects will
+/// be represetned as zero-initialized memory.
+pub const KOBJ_UNINITIALIZED: usize = 0;
+
+/// A state indicating a kernel object that is being initialized.
+pub const KOBJ_INITING: usize = 1;
+
+/// A state indicating a kernel object that has completed initialization.  This also means that the
+/// take has been called.  And shouldn't be allowed additional times.
+pub const KOBJ_INITIALIZED: usize = 2;
+
+impl<T> StaticKernelObject<T>
+where
+    StaticKernelObject<T>: Wrapped,
+{
+    /// Construct an empty of these objects, with the zephyr data zero-filled.  This is safe in the
+    /// sense that Zephyr we track the initialization, and they start in the uninitialized state.
```

Comment:
I appreciate the attention to good docs.

Not sure if these suggestions are the same as your original intent.
```suggestion
    /// Construct an empty instance of these objects, with the zephyr data zero-filled.  This is safe in the
    /// sense that in Zephyr we track the initialization, and they start in the uninitialized state.
```

Reply:
I believe this should be addressed, namely by rewriting the whole doc section.